src/rust/inetstack/protocols/tcp/queue.rs

```diff
@@ -291,7 +291,19 @@ impl<const N: usize> SharedTcpQueue<N> {
     pub async fn pop_coroutine(&mut self, size: Option<usize>, yielder: Yielder) -> Result<DemiBuffer, Fail> {
         self.state_machine.may_pop()?;
         match self.socket {
-            Socket::Established(ref mut socket) => socket.pop(size, yielder).await,
+            Socket::Established(ref mut socket) => {
+                let buf: DemiBuffer = socket.pop(size, yielder).await?;
+                // If this is an EOF on the connection, then close the socket.
+                // TODO: Get rid of this when we deduplicate the state machine.
+                // FIXME:
+                if buf.is_empty() {
+                    self.state_machine.prepare(SocketOp::RemoteClose)?;
+                    self.state_machine.commit();
+                }
+                // We are assuming that the application will call close when it receives a zero-length pop. Any other
+                // pops will be cancelled with an EBADF.
+                Ok(buf)
+            },
```

Comment:
This will work for for the happy "remote close" case. For others, we still need to modify `SharedControlBlock::receive()`